File: v3/client/LedController3Client/Ui/ColorsInterpolator.cs
using SkiaSharp;
using System;

namespace LedController3Client.Ui
{
    public class ColorsInterpolator
    {
        public void InterpolateColors(SKColor lColor, SKColor rColor, float ratio, out SKColor outColor)
        {
            var lColorMax = Math.Max(Math.Max(lColor.Red, lColor.Green), lColor.Blue);
            var rColorMax = Math.Max(Math.Max(rColor.Red, rColor.Green), rColor.Blue);
            var cColorMax = (lColorMax + rColorMax) / 2;

            var cColor = new SKColor(
                (byte)((lColor.Red + rColor.Red) / 2),
                (byte)((lColor.Green + rColor.Green) / 2),
                (byte)((lColor.Blue + rColor.Blue) / 2));

            var colorMax = Math.Max(Math.Max(cColor.Red, cColor.Green), cColor.Blue);
            if (colorMax > 0)
            {
                var mr = cColorMax / (float)colorMax;
                cColor = new SKColor(
                    (byte)(mr * cColor.Red),
                    (byte)(mr * cColor.Green),
                    (byte)(mr * cColor.Blue));
            }

            SKColor iColor;
            if (ratio < .5f)
            {
                iColor = lColor;
            }
            else
            {
                iColor = rColor;
                ratio = 1f - ratio;
            }

            ratio *= 2f;

            InterpolateColorsComponents(iColor.Red, cColor.Red, ratio, out byte r);
            InterpolateColorsComponents(iColor.Green, cColor.Green, ratio, out byte g);
            InterpolateColorsComponents(iColor.Blue, cColor.Blue, ratio, out byte b);

            outColor = new SKColor(r, g, b);
        }

        private void InterpolateColorsComponents(byte lColorComponent, byte rColorComponent, float ratio, out byte outColorComponent)
        {
            outColorComponent = (byte)(lColorComponent * (1f - ratio) + rColorComponent * ratio);
        }
    }
}

File: v3/micro/src/ColorsInterpolator.cpp
#include "ColorsInterpolator.h"
#include "Particle.h"

void ColorsInterpolator::InterpolateColors(Color_t lColor, Color_t rColor, float ratio, Color_t& outColor)
{
  uint8_t lColorMax = lColor.MaxComp();
  uint8_t rColorMax = rColor.MaxComp();
  uint8_t cColorMax = (lColorMax + rColorMax) / 2;

  Color_t cColor = Color_t(
    (uint8_t)((lColor.R + rColor.R) / 2),
    (uint8_t)((lColor.G + rColor.G) / 2),
    (uint8_t)((lColor.B + rColor.B) / 2));

  uint8_t colorMax = cColor.MaxComp();
  if (colorMax > 0)
  {
    float mr = cColorMax / (float)colorMax;
    cColor = Color_t(
      (uint8_t)(mr * cColor.R),
      (uint8_t)(mr * cColor.G),
      (uint8_t)(mr * cColor.B));
  }

  Color_t iColor;
  if (ratio < 0.5)
  {
    iColor = lColor;
  }
  else
  {
    iColor = rColor;
    ratio = 1 - ratio;
  }

  ratio *= 2;

  InterpolateColorsComponents(iColor.R, cColor.R, ratio, outColor.R);
  InterpolateColorsComponents(iColor.G, cColor.G, ratio, outColor.G);
  InterpolateColorsComponents(iColor.B, cColor.B, ratio, outColor.B);
}

void ColorsInterpolator::InterpolateColorsComponents(uint8_t lColorComponent, uint8_t rColorComponent, float ratio, uint8_t& outColorComponent)
{
  outColorComponent = (uint8_t)((float)lColorComponent * (1.0 - ratio) + (float)rColorComponent * ratio);
}
